chainer/distribution.py

```diff
@@ -0,0 +1,420 @@
+import copy
+
+
+class Distribution(object):
+
+    """Interface of Distribution
+
+    `Distribution` is a bass class to treat probability distributions.
+
+    This class provides a mean to perform following operations.
+    1. Sampling random points.
+    2. Evaluating a functions about probability at given realization values.
+        (e.g., probability density function, probability mass function)
+    3. Obtaining properties of distributions.
+        (e.g., mean, variance)
+
+    Note that every method and property that computes them from
+    `chainer.Variable` can basically be differentiated.
+
+    In this class, sampled random points and realization values given in
+    function about probability is called "sample".  Sample consists of batches,
+    and each batch consists of independent events. Each event consists of
+    values, and each value in an event cannot be sampled independently in
+    general. Each event in a batch is independent while it is not sampled from
+    an identical distribution. And each batch in sample is sampled from an
+    identical distribution.
+
+    Each part of the sample-batch-event hierarchy has its own shape, which is
+    called `sample_shape`, `batch_shape`, and `event_shape`, respectively.
+
+    On initialization, it takes distribution-specific parameters as inputs.
+    `batch_shape` and `event_shape` is decided by the shape of the parameter
+    when generating an instance of a class.
+
+    .. admonition:: Example
+
+        The following code is a example of sample-batch-event hierarchy on
+        using `MultivariateNormal` distribution. This make 2d normal
+        distributions. `dist` consists of 12(4 * 3) independent 2d normal
+        distributions. And on initialization, `batch_shape` and `event_shape`
+        is decided.
+
+        >>> import chainer
+        >>> import chainer.distributions as D
+        >>> import numpy as np
+        >>> d = 2
+        >>> shape = (4, 3)
+        >>> loc = np.random.normal(
+        ...     size=shape + (d,)).astype(np.float32)
+        >>> cov = np.random.normal(size=shape + (d, d)).astype(np.float32)
+        >>> cov = np.matmul(cov, np.rollaxis(cov, -1, -2))
+        >>> l = np.linalg.cholesky(cov)
+        >>> dist = D.MultivariateNormal(loc, l)
+        >>> dist.event_shape
+        (2,)
+        >>> dist.batch_shape
+        (4, 3)
+        >>> sample = dist.sample(sample_shape=(6, 5))
+        >>> sample.shape
+        (6, 5, 4, 3, 2)
+
+    Every function about probability takes realization value whose shape is
+    `(sample_shape, batch_shape, event_shape)` and returns an evaluated value
```

Comment:
Maybe it's better to write "the concatenation of `sample_shape`, `batch_shape`, and `event_shape`" rather than `(sample_shape, batch_shape, event_shape)` because the latter is incorrect as a Python expression and thus confusing.